File: FarmaciasDeGuardia/Services/PDFProcessingService.swift
import Foundation
import PDFKit

public class PDFProcessingService {
    /// Current active region. For now, always Segovia Capital.
    private var region: Region
    
    /// Registry of parsing strategies for each region
    private var parsingStrategies: [String: PDFParsingStrategy] = [:]
    
    public init(region: Region = .segoviaCapital) {
        self.region = region
        registerDefaultParsers()
    }
    
    /// Registers the default parsing strategies
    private func registerDefaultParsers() {
        // Register Segovia Capital parser
        register(parser: SegoviaCapitalParser(), for: "segovia-capital")
        // Register Cuellar parser
        register(parser: CuellarParser(), for: "cuellar")
        // Register El Espinar parser
        register(parser: ElEspinarParser(), for: "el-espinar")
        // Register Segovia Rural parser
        register(parser: SegoviaRuralParser(), for: "segovia-rural")
    }
    
    /// Registers a parsing strategy for a specific region
    /// - Parameters:
    ///   - parser: The parser to register
    ///   - regionId: The region identifier this parser handles
    public func register(parser: PDFParsingStrategy, for regionId: String) {
        parsingStrategies[regionId] = parser
    }
    
    /// Loads pharmacy schedules for the current region
    public func loadPharmacies(forceRefresh: Bool = false) async -> [PharmacySchedule] {
        // Get the effective PDF URL (cached or remote, force download if requested)
        let effectiveURL: URL
        if forceRefresh {
            do {
                effectiveURL = try await PDFCacheManager.shared.forceDownload(for: region)
                DebugConfig.debugPrint("✅ Force downloaded fresh PDF for \(region.name)")
            } catch {
                DebugConfig.debugPrint("❌ Failed to force download PDF for \(region.name): \(error)")
                // Fallback to regular effective URL
                effectiveURL = await region.getEffectivePDFURL()
            }
        } else {
            effectiveURL = await region.getEffectivePDFURL()
        }
        
        guard let pdfDocument = PDFDocument(url: effectiveURL) else {
            DebugConfig.debugPrint("Failed to load PDF from \(effectiveURL)")
            return []
        }

        guard let parser = parsingStrategies[region.id] else {
            DebugConfig.debugPrint("No parser found for region: \(region.name) (id: \(region.id))")
            return []
        }

        DebugConfig.debugPrint("Loading schedules for \(region.name) from \(effectiveURL)")
        return parser.parseSchedules(from: pdfDocument)
    }

    /// Updates the current region and returns schedules for that region
    /// - Parameter newRegion: The new region to update to
    /// - Parameter forceRefresh: Whether to force re-download the PDF
    /// - Returns: An array of `PharmacySchedule` for the new region
    public func loadPharmacies(for newRegion: Region, forceRefresh: Bool = false) async -> [PharmacySchedule] {
        self.region = newRegion
        return await loadPharmacies(forceRefresh: forceRefresh)
    }    /// Internal method, kept for backward compatibility and testing
    func loadPharmacies(from url: URL) -> [PharmacySchedule] {
        guard let pdfDocument = PDFDocument(url: url) else {
            DebugConfig.debugPrint("Failed to load PDF from \(url)")
            return []
        }
        
        // For direct URL loading, use Segovia parser as default
        let parser = parsingStrategies["segovia-capital"] ?? SegoviaCapitalParser()
        DebugConfig.debugPrint("Loading schedules from \(url)")
        return parser.parseSchedules(from: pdfDocument)
    }
}

File: android/app/src/main/java/com/github/bfollon/farmaciasdeguardiaensegovia/services/PDFProcessingService.kt
/*
 * Copyright (C) 2025  Bruno Follon (@bFollon)
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package com.github.bfollon.farmaciasdeguardiaensegovia.services

import com.github.bfollon.farmaciasdeguardiaensegovia.data.DutyLocation
import com.github.bfollon.farmaciasdeguardiaensegovia.data.PharmacySchedule
import com.github.bfollon.farmaciasdeguardiaensegovia.data.Region
import com.github.bfollon.farmaciasdeguardiaensegovia.services.pdfparsing.PDFParsingStrategy
import com.github.bfollon.farmaciasdeguardiaensegovia.services.pdfparsing.strategies.CuellarParser
import com.github.bfollon.farmaciasdeguardiaensegovia.services.pdfparsing.strategies.ElEspinarParser
import com.github.bfollon.farmaciasdeguardiaensegovia.services.pdfparsing.strategies.SegoviaCapitalParser
import com.github.bfollon.farmaciasdeguardiaensegovia.services.pdfparsing.strategies.SegoviaRuralParser
import java.io.File
import kotlin.collections.component1
import kotlin.collections.component2

/**
 * Android equivalent of iOS PDFProcessingService
 * Handles parsing PDF files to extract pharmacy schedule information using PDFBox
 */
class PDFProcessingService {

    // Registry of parsing strategies for each region
    private val parsingStrategies: Map<String, PDFParsingStrategy> = mapOf(
        "segovia-capital" to SegoviaCapitalParser(),
        "cuellar" to CuellarParser(),
        "el-espinar" to ElEspinarParser(),
        "segovia-rural" to SegoviaRuralParser()
    )

    /**
     * Load pharmacy schedules from a PDF file
     * Equivalent to iOS PDFProcessingService.loadPharmacies
     *
     * @param pdfFile The PDF file to process
     * @param region The region this PDF belongs to
     * @param forceRefresh Whether to bypass any internal caching
     * @return List of pharmacy schedules extracted from the PDF
     */
    suspend fun loadPharmacies(
        pdfFile: File,
        region: Region
    ): Map<DutyLocation, List<PharmacySchedule>> {
        DebugConfig.debugPrint("\n=== PDF Processing Started ===")
        DebugConfig.debugPrint("PDFProcessingService: Processing PDF for ${region.name}")
        DebugConfig.debugPrint("📁 PDF file: ${pdfFile.absolutePath}")
        DebugConfig.debugPrint("📏 File size: ${pdfFile.length()} bytes")
        DebugConfig.debugPrint("📄 File exists: ${pdfFile.exists()}")
        DebugConfig.debugPrint("🔄 Force refresh: ${region.forceRefresh}")

        if (!pdfFile.exists() || pdfFile.length() == 0L) {
            DebugConfig.debugError("PDFProcessingService: PDF file does not exist or is empty")
            return emptyMap()
        }

        return try {
            // Get the appropriate parsing strategy for this region
            val parsingStrategy = getParsingStrategy(region)
            DebugConfig.debugPrint("📋 Using parsing strategy: ${parsingStrategy::class.simpleName}")

            // Parse the schedules using the strategy
            DebugConfig.debugPrint("⚙️ Starting PDF parsing...")
            val scheduleMap = parsingStrategy.parseSchedules(pdfFile)

            scheduleMap.forEach { (location, schedules) ->
                DebugConfig.debugPrint("✅ PDFProcessingService: Successfully parsed ${schedules.size} schedules from ${location.name} PDF")
            }

            // Log some sample data for debugging
            if (scheduleMap.isEmpty()) {
                DebugConfig.debugWarn("⚠️ No schedules were parsed from the PDF!")
            }

            DebugConfig.debugPrint("=== PDF Processing Complete ===\n")
            scheduleMap
        } catch (e: Exception) {
            DebugConfig.debugError(
                "❌ PDFProcessingService: Error processing PDF for ${region.name}: ${e.message}",
                e
            )
            emptyMap()
        }
    }

    /**
     * Get the appropriate parsing strategy for a region
     */
    private fun getParsingStrategy(region: Region): PDFParsingStrategy {
        return parsingStrategies[region.id] ?: SegoviaCapitalParser()
    }

}
